doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,706 @@
+   be defined on the class as is normal in Python.
+   Storage information such as itemsize and byteorder can differ between
+   different dtype instances (e.g. "S3" vs. "S8") and will remain part of the instance.
+   This means that in the long run the current lowlevel access to dtype methods
+   will be removed (see ``ArrFunctions`` in NEP 40).
+
+2. The current NumPy scalars will *not* change, they will not be instances of
+   datatypes. This will also be true for new datatypes, scalars will not be
+   instances of a dtype (although ``isinstance(scalar, dtype)`` may be defined).
+
+With detailed technical decisions being followed up in NEP 42.
+
+Further, the public API will be designed in a way that is extensible in the future:
+
+3. All new API functions provided to the user will hide implementation details
+   as much as possible. The public API should be an identical, but limited,
+   version of the API used for the internal NumPy datatypes.
+
+The changes to the datatype system in Phase II must include a large refactor of the
+UFunc machinery, which will be further defined in NEP 43:
+
+4. To enable all of the desired functionality for new user defined datatypes,
+   the UFunc machinery will be changed to replace the current dispatching
+   and type resolution system.
+   The old system should be *mostly* supported as a legacy version for some time.
+
+Additionally, as a general design principle, the addition of new user defined
+datatypes will *not* change the behaviour of programs.
+For example ``common_dtype(a, b)`` must not be ``c`` unless ``a`` or ``b`` know
+that ``c`` exists.
+
+
+User Impact
+-----------
+
+The current ecosystem has very few user defined datatypes using NumPy, the
+two most prominent being: ``rational`` and ``quaternion``.
+These represent fairly simple datatypes which are not strongly impacted
+by the current limitations.
+However, we have identified a need for datatypes such as:
+
+* bfloat16, used in deep learning
+* categorical types
+* physical units (such as meters)
+* datatypes for tracing/automatic differentiation
+* high, fixed precision math
+* specifialized integer types such as int2, int24
+* new, better datetime representations
+* extending e.g. integer dtypes to have a sentinel NA value
+* geometrical objects [pygeos]_
+
+Some of these are partially solved; for example unit capability is provided
+in ``astropy.units``, ``unyt``, or ``pint``, as `numpy.ndarray` subclasses.
+Most of these datatypes, however, simply cannot be reasonably defined
+right now.
+An advantage of having such datatypes in NumPy is that they should integrate
+seamlessly with other array or array-like packages such as Pandas,
+``xarray``, or ``Dask``.
+
+The long term user impact of implementing this NEP will be to allow both
+the growth of the whole ecosystem by having such new datatypes, as well as
+consolidating implementation of such datatypes within NumPy to achieve
+a better interoperability.
+
+
+Examples
+^^^^^^^^
```

Comment:
Perhaps worth a disclaimer saying that the exact behavior of these examples (eg the choices of casting rules) are not in the scope of this NEP - merely they're an example of the type of choice we want to allow users to be able to make.